lib/scan-includes.js

```diff
@@ -0,0 +1,110 @@
+    }
+
+    /** All the files that successfully fetched go here. */
+    const recursiveIncludes = new Set();
+    /** When this gets back to empty, we're done searching includes. */
+    const inProgressFetches = new Set();
+
+    await new Promise((resolve, reject) => {
+        async function scanOneFile(path) {
+            try {
+                console.log("Scanning", path);
+                if (path.startsWith('/') ||
+                    path.includes('../') ||
+                    // Block schemes, a bit more aggressively than necessary.
+                    path.includes(':')) {
+                    return;
+                }
+                let resolvedUrl;
+                try {
+                    resolvedUrl = new URL(path, repositoryUrl).href;
+                    if (!resolvedUrl.startsWith(repositoryUrl)) {
+                        console.error(resolvedUrl, "doesn't start with", repositoryUrl);
+                        // Can't include paths outside the repository.
+                        return;
+                    }
+                } catch (e) {
+                    console.error(e);
+                    // Don't treat broken URLs as includes.
+                    return;
+                }
+
+                inProgressFetches.add(path);
+                recursiveIncludes.add(path);
+                try {
+                    console.log("Fetching", resolvedUrl);
+                    body = await fetchUrl(resolvedUrl, GITHUB_SERVICE);
+                    console.log("Got", resolvedUrl);
+                    for (const match of body.matchAll(includeRE)) {
+                        for (let i = 1; i < match.length; i++) {
+                            const included = match[i].trim();
+                            if (included) {
+                                console.log("Found include:", included);
+                                scanOneFile(included).catch(reject);
+                            }
+                        }
+                    }
+                } catch (err) {
+                    console.error(resolvedUrl, "failed with", err);
+                    // This include scan can have false positives, so if one fails to fetch,
+                    // we assume it's not a real include.
+                    recursiveIncludes.delete(path);
+                }
+            } finally {
+                inProgressFetches.delete(path);
+                if (inProgressFetches.size == 0) {
+                    resolve();
+                }
+            }
+        }
+
+        scanOneFile(rootUrl).catch(reject);
+    });
+
+    return recursiveIncludes;
+}
+
+/** If `nestedUrl` is under `baseUrl`, returns the relative URL such that `new
+ * URL(relative, baseUrl)` returns `nestedUrl`.
+ *
+ * Otherwise throws an Error.
+ *
+ * @param {string} baseUrl
+ * @param {string} nestedUrl
+ */
+function enforceRelative(baseUrl, nestedUrl) {
```

Comment:
This seems to be unused.

Reply:
Yep, I hadn't adequately cleaned up the PR.